File: app/src/main/java/com/qiming/kurtapp/ui/MarqueeTextView.java
package com.qiming.kurtapp.ui;

import android.content.Context;
import android.support.v7.widget.AppCompatTextView;
import android.text.TextUtils;
import android.util.AttributeSet;

/**
 * Created by kurtg on 18/2/3.
 */

public class MarqueeTextView extends AppCompatTextView {

    private boolean focus;

    public MarqueeTextView(Context context) {
        super(context);
        init();
    }

    public MarqueeTextView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public MarqueeTextView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        setSingleLine();
        setEllipsize(TextUtils.TruncateAt.MARQUEE);
        setMarqueeRepeatLimit(-1);
//        setMaxEms(10);
    }

    @Override
    public boolean isFocused() {
        return focus;
    }

    public void setFocus(final boolean b) {
        focus = b;
        post(new Runnable() {
            @Override
            public void run() {
                MarqueeTextView.super.setSelected(b);
            }
        });
    }
}

File: app/src/main/java/com/kurt/widget/MarqueeTextView.kt
package com.kurt.widget

import android.content.Context
import android.support.v7.widget.AppCompatTextView
import android.text.TextUtils
import android.util.AttributeSet

class MarqueeTextView: AppCompatTextView {
    var focus: Boolean = false

    constructor(context: Context) : super(context)
    constructor(context: Context, aSet: AttributeSet?) : super(context, aSet)

    init {
        setSingleLine()
        ellipsize = TextUtils.TruncateAt.MARQUEE
        marqueeRepeatLimit = -1
    }

    override fun isFocused(): Boolean {
        return focus
    }

    fun marquee(b: Boolean) {
        focus = b
        post { super@MarqueeTextView.setSelected(b) }
    }
}